Core/src/RequestWrapper.php

```diff
@@ -309,54 +311,102 @@ private function applyHeaders(RequestInterface $request, array $options = [])
             $quotaProject = $this->quotaProject;
             $token = null;
 
+            $authHeaders = [];
             if ($this->accessToken) {
-                $token = $this->accessToken;
+                $authHeaders = $this->createAuthHeader($this->accessToken);
             } else {
                 $credentialsFetcher = $this->getCredentialsFetcher();
-                $token = $this->fetchCredentials($credentialsFetcher)['access_token'];
+                $authHeaders = $this->fetchAuthHeaders($credentialsFetcher);
 
                 if ($credentialsFetcher instanceof GetQuotaProjectInterface) {
                     $quotaProject = $credentialsFetcher->getQuotaProject();
                 }
             }
 
-            $headers['Authorization'] = 'Bearer ' . $token;
-
             if ($quotaProject) {
                 $headers['X-Goog-User-Project'] = [$quotaProject];
             }
+
+            foreach ($authHeaders as $key => $value) {
+                if (array_key_exists($key, $headers)) {
+                    $headers[$key] .= ' ' . $value;
+                } else {
+                    $headers[$key] = $value;
+                }
+            }
         }
 
         return Utils::modifyRequest($request, ['set_headers' => $headers]);
     }
 
     /**
-     * Fetches credentials.
+     * Fetches auth headers.
      *
      * @param FetchAuthTokenInterface $credentialsFetcher
      * @return array
      * @throws ServiceException
      */
-    private function fetchCredentials(FetchAuthTokenInterface $credentialsFetcher)
+    private function fetchAuthHeaders(FetchAuthTokenInterface $credentialsFetcher)
     {
         $backoff = new ExponentialBackoff($this->retries, $this->getRetryFunction());
 
         try {
             return $backoff->execute(
                 function () use ($credentialsFetcher) {
-                    if ($token = $credentialsFetcher->fetchAuthToken($this->authHttpHandler)) {
-                        return $token;
+                    // Proper error message capture to as to not break existing
+                    // exception message.
+                    $exceptionMessage = '';
+
+                    // We need to find the actual fetcher incase of a cache wrapper
+                    // so that we can avoid the exception case where actual fetcher
+                    // does not implements UpdateMetadataInterface with cache wrapper's
+                    // `updateMetadata` being called.
+                    $actualFetcher = $credentialsFetcher;
+                    if ($credentialsFetcher instanceof FetchAuthTokenCache) {
+                        $actualFetcher = $credentialsFetcher->getFetcher();
+                    }
+
+                    if ($actualFetcher instanceof UpdateMetadataInterface) {
```

Comment:
We can simplify this as we did in the auth library:
```suggestion
                    if ($this->fetcher instanceof UpdateMetadataInterface &&
                       !($this->fetcher FetchAuthTokenCache && !$this->fetcher->getFetcher() instanceof UpdateMetadataInterface)
                    ) {
```

Reply:
Done.